### PR TITLE
feat: support `?module` imports for WebAssembly

### PR DESCRIPTION
**Pull request by ForsakenHarmony**

### CODE REVIEW

**Comment by ForsakenHarmony** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5618** <a href="https://app.graphite.dev/github/pr/vercel/turbo/5618" target="_blank">Graphite</a> 
    * **PR #5636** <a href="https://app.graphite.dev/github/pr/vercel/turbo/5636" target="_blank">Graphite</a> 
      * **PR #5660** <a href="https://app.graphite.dev/github/pr/vercel/turbo/5660" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/vercel/turbo/5660?utm_source=stack-comment).

File: crates/turbopack-ecmascript-runtime/js/src/dev/runtime/dom/runtime-backend-dom.ts [hunk 1]
```diff
@@ -40,6 +40,17 @@ async function loadWebAssembly(
...
 
+async function loadWebAssemblyModule(
+  _source: SourceInfo,
+  wasmChunkPath: ChunkPath
+): Promise<WebAssembly.Module> {
+  const chunkUrl = `/${getChunkRelativeUrl(wasmChunkPath)}`;
+
+  const req = fetch(chunkUrl);
+
+  return WebAssembly.compileStreaming(req);
```
Comment (jridgewell) [inline crates/turbopack-ecmascript-runtime/js/src/dev/runtime/dom/runtime-backend-dom.ts:51]: Nit: there's a small perf and dx improvement by awaiting the promise before returning
```suggestion
  return await WebAssembly.compileStreaming(req);
```

File: crates/turbopack-ecmascript-runtime/js/src/dev/runtime/dom/runtime-backend-dom.ts [hunk 2]
```diff
@@ -40,6 +40,17 @@ async function loadWebAssembly(
   return instance.exports;
```
Comment (jridgewell) [inline crates/turbopack-ecmascript-runtime/js/src/dev/runtime/dom/runtime-backend-dom.ts:40]: Can `loadWebAssembly` be rewritten to use `loadWebAssemblyModule`?

File: crates/turbopack-ecmascript-runtime/js/src/shared-node/node-utils.ts [hunk 3]
```diff
@@ -47,7 +48,17 @@ externalRequire.resolve = (
...
 
-async function loadWebAssemblyFromPath(
+async function compileWebAssemblyFromPath(
+  path: string
+): Promise<WebAssembly.Module> {
+  const { readFile } = require("fs/promises") as typeof import("fs/promises");
+
+  const buffer = await readFile(path);
+
+  return WebAssembly.compile(buffer);
```
Comment (jridgewell) [inline crates/turbopack-ecmascript-runtime/js/src/shared-node/node-utils.ts:58]: Ditto
```suggestion
  return await WebAssembly.compile(buffer);
```

File: crates/turbopack-ecmascript-runtime/js/src/shared-node/node-utils.ts [hunk 4]
```diff
@@ -47,7 +48,17 @@ externalRequire.resolve = (
...
+}
+
+async function instantiateWebAssemblyFromPath(
```
Comment (jridgewell) [inline crates/turbopack-ecmascript-runtime/js/src/shared-node/node-utils.ts:61]: Ditto reusing `compileWebAssemblyFromPath`

File: crates/turbopack-ecmascript-runtime/js/src/shared-node/node-utils.ts [hunk 5]
```diff
@@ -47,7 +48,17 @@ externalRequire.resolve = (
   return require.resolve(id, options);
 };
 
-async function loadWebAssemblyFromPath(
+async function compileWebAssemblyFromPath(
+  path: string
+): Promise<WebAssembly.Module> {
+  const { readFile } = require("fs/promises") as typeof import("fs/promises");
+
+  const buffer = await readFile(path);
```
Comment (jridgewell) [inline crates/turbopack-ecmascript-runtime/js/src/shared-node/node-utils.ts:56]: Nit: Can we use a streaming file read instead of a full buffer?

**Review by jridgewell — APPROVED** [review]

File: crates/turbopack-ecmascript-runtime/js/src/shared-node/node-utils.ts (hunk 5 above)
Comment (ForsakenHarmony) [reply crates/turbopack-ecmascript-runtime/js/src/shared-node/node-utils.ts:56]: Is that possible? WebAssembly seems to only supports a request (mirroring the Web api), not a file stream

Comment (jridgewell) [reply crates/turbopack-ecmascript-runtime/js/src/shared-node/node-utils.ts]: Yup, just need to create a response that wraps the stream:

```js
const stream = createReadStream("path/to/module.wasm");
const response = new Response(Readable.toWeb(stream), {
  headers: {
    "content-type": "application/wasm",
  },
});
const { instance } = await WebAssembly.instantiateStreaming(response);
```

Comment (ForsakenHarmony) [reply crates/turbopack-ecmascript-runtime/js/src/shared-node/node-utils.ts]: Ah nice, will use that

**Review by jridgewell — APPROVED** [review]

**Pull request merged**